sbin/mount/mount.8
.\" Copyright (c) 1980, 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)mount.8	8.8 (Berkeley) 6/16/94
.\" $FreeBSD: src/sbin/mount/mount.8,v 1.31.2.12 2003/02/23 21:17:42 trhodes Exp $
.\"
.Dd May 15, 2016
.Dt MOUNT 8
.Os
.Sh NAME
.Nm mount
.Nd mount file systems
.Sh SYNOPSIS
.Nm
.Op Fl adfpruvw
.Op Fl F Ar fstab
.Op Fl o Ar options
.Op Fl t Ar type
.Nm
.Op Fl dfpruvw
.Brq Ar special | node
.Nm
.Op Fl dfpruvw
.Op Fl o Ar options
.Op Fl t Ar type
.Ar special node
.Sh DESCRIPTION
The
.Nm
utility calls the
.Xr mount 2
system call to prepare and graft a
.Ar "special device"
or the remote node (rhost:path) or a HAMMER2 label (@label) on to the
file system tree at the point
.Ar node .
If either
.Ar special
or
.Ar node
are not provided, the appropriate information is taken from the
.Xr fstab 5
file.
.Pp
The system maintains a list of currently mounted file systems.
If no arguments are given to
.Nm ,
this list is printed.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl a
All the filesystems described in
.Xr fstab 5
are mounted.
Exceptions are those marked as
.Dq noauto ,
excluded by the
.Fl t
flag (see below), or if they are already mounted (except the
root filesystem which is always remounted to preserve
traditional single user mode behavior).
.It Fl d
Causes everything to be done except for the actual system call.
This option is useful in conjunction with the
.Fl v
flag to
determine what the
.Nm
command is trying to do.
.It Fl F Ar fstab
Specify the
.Pa fstab
file to use.
.It Fl f
Forces the revocation of write access when trying to downgrade
a filesystem mount status from read-write to read-only.
Also
forces the R/W mount of an unclean filesystem (dangerous; use with
caution).
.It Fl o
Options are specified with a
.Fl o
flag followed by a comma separated string of options.
If a
.Dq no
prefix is added or removed from an option name, then meaning is negated.
In case of conflicting options being specified, the rightmost option
takes effect.
The following options are available:
.Bl -tag -width indent
.It Cm async
All
.Tn I/O
to the file system should be done asynchronously.
This is a
.Em dangerous
flag to set,
and should not be used unless you are prepared to recreate the file
system should your system crash.
.It Cm automounted
This flag indicates that the file system was mounted by
.Xr automountd 8 .
Automounted file systems are automatically unmounted by
.Xr autounmountd 8 .
.It Cm current
When used with the
.Fl u
flag, this is the same as specifying the options currently in effect for
the mounted filesystem.
.It Cm force
The same as
.Fl f ;
forces the revocation of write access when trying to downgrade
a filesystem mount status from read-write to read-only.
Also
forces the R/W mount of an unclean filesystem (dangerous; use with caution).
.It Cm fstab
When used with the
.Fl u
flag, this is the same as specifying all the options listed in the
.Xr fstab 5
file for the filesystem.
.It Cm noasync
Metadata I/O should be done synchronously, while data I/O should be done
asynchronously.
This is the default.
.It Cm noatime
Do not update the file access time when reading from a file.
This option
is useful on filesystems where there are large numbers of files and
performance is more critical than updating the file access time (which is
rarely ever important).
This option is currently only supported on local filesystems.
.It Cm noauto
This filesystem should be skipped when
.Nm
is run with the
.Fl a
flag.
.It Cm noclusterr
Disable read clustering.
.It Cm noclusterw
Disable write clustering.
.It Cm nodev
Do not interpret character or block special devices on the file system.
This option is useful for a server that has file systems containing
special devices for architectures other than its own.
This option is set automatically when the user does not have super-user
privileges.
.It Cm noexec
Do not allow execution of any binaries on the mounted file system.
This option is useful for a server that has file systems containing
binaries for architectures other than its own.
.It Cm nosuid
Do not allow set-user-identifier or set-group-identifier bits to take effect.
Note: this option is worthless if a publicly available suid or sgid
wrapper like
.Xr suidperl 1
is installed on your system.
It is set automatically when the user does not have super-user privileges.
.It Cm nosymfollow
Do not follow symlinks
on the mounted file system.
.It Cm rdonly , ro , norw
The same as
.Fl r ;
mount the file system read-only (even the super-user may not write it).
.It Cm sync
All
.Tn I/O
to the file system should be done synchronously.
.It Cm trim
If the device supports TRIM
.Va ( kern.cam.da.X.trim_enabled
exists) and is set,
the file system will perform online trim for corresponding block deletions.
Currently, only
.Xr UFS 5
supports this feature.
.It Cm suiddir
A directory on the mounted filesystem will respond to the SUID bit
being set, by setting the owner of any new files to be the same
as the owner of the directory.
New directories will inherit the bit from their parents.
Execute bits are removed from
the file, and it will not be given to root.
.Pp
This feature is designed for use on fileservers serving PC users via
ftp or SAMBA.
It provides security holes for shell users and as
such should not be used on shell machines, especially on home directories.
This option requires the SUIDDIR
option in the kernel to work.
Only
.Xr UFS 5
filesystems support this option.
See
.Xr chmod 2
for more information.
.It Cm update
The same as
.Fl u ;
indicate that the status of an already mounted file system should be changed.
.It Cm ignore
Will be ignored by
.Xr df 1 .
.El
.Pp
Any additional options specific to a filesystem type that is not
one of the internally known types (see the
.Fl t
option) may be passed as a comma separated list; these options are
distinguished by a leading
.Dq \&-
(dash).
Options that take a value are specified using the syntax -option=value.
For example, the
.Nm
command:
.Bd -literal -offset indent
mount -t mfs -o nosuid,-N,-s=4000 /dev/da0s0b /tmp
.Ed
.Pp
causes
.Nm
to execute the equivalent of:
.Bd -literal -offset indent
/sbin/mount_mfs -o nosuid -N -s 4000 /dev/da0s0b /tmp
.Ed
.Pp
Additional options specific to filesystem types
which are not internally known
(see the description of the
.Fl t
option below)
may be described in the manual pages for the associated
.Pa /sbin/mount_ Ns Sy XXX
utilities.
.It Fl p
Print mount information in
.Xr fstab 5
format.
If fstab is missing or if the freq and passno fields are omitted,
the default values as described in
.Xr fstab 5
are used.
Implies also the
.Fl v
option.
.It Fl r
The file system is to be mounted read-only.
Mount the file system read-only (even the super-user may not write it).
The same as the
.Cm rdonly
argument to the
.Fl o
option.
.It Fl t Ar type
The argument following the
.Fl t
is used to indicate the file system type.
The type
.Cm ufs
is the default.
The
.Fl t
option can be used
to indicate that the actions should only be taken on
filesystems of the specified type.
More than one type may be specified in a comma separated list.
The list of filesystem types can be prefixed with
.Dq no
to specify the filesystem types for which action should
.Em not
be taken.
For example, the
.Nm
command:
.Bd -literal -offset indent
mount -a -t nonfs,mfs
.Ed
.Pp
mounts all filesystems except those of type
.Tn NFS
and
.Tn MFS .
.Pp
If the type is not the internally known type,
.Cm ufs ,
.Nm
will attempt to execute a program in
.Pa /sbin/mount_ Ns Sy XXX
where
.Sy XXX
is replaced by the type name.
For example, nfs filesystems are mounted by the program
.Pa /sbin/mount_nfs .
.Pp
Most filesystems will be dynamically loaded by their
.Nm
programs
if not already present in the kernel, using the
.Xr vfsload 3
subroutine.
Because this mechanism requires writable temporary space,
the filesystem type containing
.Pa /tmp
must be compiled into the kernel, and the filesystems containing
.Pa /tmp
and
.Pa /usr/bin/ld
must be listed in
.Pa /etc/fstab
before any filesystems which might be dynamically loaded.
.It Fl u
The
.Fl u
flag indicates that the status of an already mounted file
system should be changed.
Any of the options discussed above (the
.Fl o
option)
may be changed;
also a file system can be changed from read-only to read-write
or vice versa.
An attempt to change from read-write to read-only will fail if any
files on the filesystem are currently open for writing unless the
.Fl f
flag is also specified.
The set of options is determined by applying the options specified
in the argument to
.Fl o
and finally applying the
.Fl r
or
.Fl w
option.
.It Fl v
Verbose mode.
.It Fl w
The file system object is to be read and write.
.El
.Sh ENVIRONMENT
.Bl -tag -width PATH_FSTAB
.It Pa PATH_FSTAB
If the environment variable
.Pa PATH_FSTAB
is set all operations are performed against the specified file.
.El
.Sh FILES
.Bl -tag -width /etc/fstab -compact
.It Pa /etc/fstab
file system table
.El
.Sh DIAGNOSTICS
Various, most of them are self-explanatory.
.Pp
.Dl XXXXX filesystem is not available
.Pp
The kernel does not support the respective filesystem type.
Note that
support for a particular filesystem might be provided either on a static
(kernel compile-time), or dynamic basis (loaded as a kernel module by
.Xr kldload 8 ) .
Normally,
.Nm
or its subprocesses attempt to dynamically load a filesystem module if
it has not been configured statically, using
.Xr vfsload 3 .
In this case, the above error message can also mean that you did not
have permission to load the module.
.Sh SEE ALSO
.Xr df 1 ,
.Xr lsvfs 1 ,
.Xr mount 2 ,
.Xr vfsload 3 ,
.Xr devtab 5 ,
.Xr fstab 5 ,
.Xr UFS 5 ,
.Xr automount 8 ,
.Xr automountd 8 ,
.Xr autounmountd 8 ,
.Xr kldload 8 ,
.Xr mount_cd9660 8 ,
.Xr mount_devfs 8 ,
.Xr mount_ext2fs 8 ,
.Xr mount_hammer 8 ,
.Xr mount_hammer2 8 ,
.Xr mount_hpfs 8 ,
.Xr mount_mfs 8 ,
.Xr mount_msdos 8 ,
.Xr mount_nfs 8 ,
.Xr mount_ntfs 8 ,
.Xr mount_null 8 ,
.Xr mount_procfs 8 ,
.Xr mount_smbfs 8 ,
.Xr mount_std 8 ,
.Xr mount_tmpfs 8 ,
.Xr mount_udf 8 ,
.Xr sysctl 8 ,
.Xr umount 8
.Sh HISTORY
A
.Nm
utility appeared in
.At v1 .
.Sh CAVEATS
After a successful
.Nm ,
the permissions on the original mount point determine if
.Pa ..\&
is accessible from the mounted file system.
The minimum permissions for
the mount point for traversal across the mount point in both
directions to be possible for all users is 0111 (execute for all).
.Sh BUGS
It is possible for a corrupted file system to cause a crash.


usr.bin/dsynth/mount.c
/*
 * Copyright (c) 2019 The DragonFly Project.  All rights reserved.
 *
 * This code is derived from software contributed to The DragonFly Project
 * by Matthew Dillon <dillon@backplane.com>
 *
 * This code uses concepts and configuration based on 'synth', by
 * John R. Marino <draco@marino.st>, which was written in ada.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 * 3. Neither the name of The DragonFly Project nor the names of its
 *    contributors may be used to endorse or promote products derived
 *    from this software without specific, prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
 * COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
#include "dsynth.h"

static void domount(worker_t *work, int type,
			const char *spath, const char *dpath,
			const char *discretefmt);
static void dounmount(worker_t *work, const char *rpath);
static void makeDiscreteCopies(const char *spath, const char *discretefmt);

/*
 * Called by the frontend to create a template which will be cpdup'd
 * into fresh workers.
 *
 * Template must have been previously destroyed.  Errors are fatal
 */
int
DoCreateTemplate(int force)
{
	struct stat st;
	char *goodbuf;
	char *buf;
	int rc;
	int fd;
	int n;

	rc = 0;
	asprintf(&goodbuf, "%s/.template.good", BuildBase);

	/*
	 * Conditionally create the template and discrete copies of certain
	 * directories if we think we are missing things.
	 */
	if (force == 0) {
		time_t ls_mtime;

		asprintf(&buf, "%s/Template", BuildBase);
		if (stat(buf, &st) < 0)
			force = 1;
		free(buf);

		/*
		 * Check to see if the worker count changed and some
		 * template dirs are missing or out of date, and also if
		 * a new world was installed (via /bin/ls mtime).
		 */
		asprintf(&buf, "%s/bin/ls", SystemPath);
		if (stat(buf, &st) < 0)
			dfatal_errno("Unable to locate %s", buf);
		free(buf);
		ls_mtime = st.st_mtime;

		for (n = 0; n < MaxWorkers; ++n) {
			asprintf(&buf, "%s/bin.%03d/ls", BuildBase, n);
			if (stat(buf, &st) < 0 || st.st_mtime != ls_mtime)
				force = 1;
			free(buf);
		}

		if (stat(goodbuf, &st) < 0)
			force = 1;
	}

	dlog(DLOG_ALL, "Check Template: %s\n",
	     (force ? "Must-Create" : "Good"));

	/*
	 * Create the template
	 */
	if (force) {
		remove(goodbuf);	/* ignore exit code */

		rc = 0;
		asprintf(&buf, "%s/mktemplate %s %s/Template",
			 SCRIPTPATH(SCRIPTDIR), SystemPath, BuildBase);
		rc = system(buf);
		if (rc)
			dfatal("Command failed: %s\n", buf);
		dlog(DLOG_ALL | DLOG_FILTER,
		     "Template - rc=%d running %s\n", rc, buf);
		free(buf);

		/*
		 * Make discrete copies of certain extremely heavily used
		 * but small directories.
		 */
		makeDiscreteCopies("$/bin", "/bin.%03d");
		makeDiscreteCopies("$/lib", "/lib.%03d");
		makeDiscreteCopies("$/libexec", "/libexec.%03d");
		makeDiscreteCopies("$/usr/bin", "/usr.bin.%03d");

		/*
		 * Mark the template good... ah, do a sync() to really
		 * be sure that it can't get corrupted.
		 */
		sync();
		fd = open(goodbuf, O_RDWR|O_CREAT|O_TRUNC, 0644);
		dassert_errno(fd >= 0, "could not create %s", goodbuf);
		close(fd);

		dlog(DLOG_ALL | DLOG_FILTER, "Template - done\n");
	}
	free(goodbuf);

	return force;
}

void
DoDestroyTemplate(void)
{
	struct stat st;
	char *path;
	char *buf;
	int rc;

	/*
	 * NOTE: rm -rf safety, use a fixed name 'Template' to ensure we
	 *	 do not accidently blow something up.
	 */
	asprintf(&path, "%s/Template", BuildBase);
	if (stat(path, &st) == 0) {
		asprintf(&buf, "chflags -R noschg %s; /bin/rm -rf %s",
			 path, path);
		rc = system(buf);
		if (rc)
			dfatal("Command failed: %s (ignored)\n", buf);
		free(buf);
	}
	free(path);
}

/*
 * Called by the worker support thread to install a new worker
 * filesystem topology.
 */
void
DoWorkerMounts(worker_t *work)
{
	char *buf;
	int rc;

	/*
	 * Generate required mounts, domount() will mkdir() the target
	 * directory if necessary and prefix spath with SystemPath if
	 * it starts with $/
	 */
	domount(work, TMPFS_RW, "dummy", "", NULL);
	asprintf(&buf, "%s/usr", work->basedir);
	if (mkdir(buf, 0755) != 0) {
		fprintf(stderr, "Command failed: mkdir %s\n", buf);
		++work->mount_error;
	}
	domount(work, NULLFS_RO, "$/boot", "/boot", NULL);
	domount(work, TMPFS_RW,  "dummy", "/boot/modules.local", NULL);
	domount(work, DEVFS_RW,  "dummy", "/dev", NULL);
	domount(work, PROCFS_RO, "dummy", "/proc", NULL);
	domount(work, NULLFS_RO, "$/bin", "/bin", "/bin.%03d");
	domount(work, NULLFS_RO, "$/sbin", "/sbin", NULL);
	domount(work, NULLFS_RO, "$/lib", "/lib", "/lib.%03d");
	domount(work, NULLFS_RO, "$/libexec", "/libexec", "/libexec.%03d");
	domount(work, NULLFS_RO, "$/usr/bin", "/usr/bin", "/usr.bin.%03d");
	domount(work, NULLFS_RO, "$/usr/include", "/usr/include", NULL);
	domount(work, NULLFS_RO, "$/usr/lib", "/usr/lib", NULL);
	domount(work, NULLFS_RO, "$/usr/libdata", "/usr/libdata", NULL);
	domount(work, NULLFS_RO, "$/usr/libexec", "/usr/libexec", NULL);
	domount(work, NULLFS_RO, "$/usr/sbin", "/usr/sbin", NULL);
	domount(work, NULLFS_RO, "$/usr/share", "/usr/share", NULL);
	domount(work, TMPFS_RW_MED,  "dummy", "/usr/local", NULL);
	domount(work, NULLFS_RO, "$/usr/games", "/usr/games", NULL);
	if (UseUsrSrc)
		domount(work, NULLFS_RO, "$/usr/src", "/usr/src", NULL);
	domount(work, NULLFS_RO, DPortsPath, "/xports", NULL);
	domount(work, NULLFS_RW, OptionsPath, "/options", NULL);
	domount(work, NULLFS_RW, PackagesPath, "/packages", NULL);
	domount(work, NULLFS_RW, DistFilesPath, "/distfiles", NULL);
	domount(work, TMPFS_RW_BIG, "dummy", "/construction", NULL);
	if (UseCCache)
		domount(work, NULLFS_RW, CCachePath, "/ccache", NULL);

	/*
	 * NOTE: Uses blah/. to prevent cp from creating 'Template' under
	 *	 work->basedir.  We want to start with the content.
	 */
	asprintf(&buf, "cp -Rp %s/Template/. %s", BuildBase, work->basedir);
	rc = system(buf);
	if (rc) {
		fprintf(stderr, "Command failed: %s\n", buf);
		++work->accum_error;
		snprintf(work->status, sizeof(work->status),
			 "Template copy failed");
	}
	free(buf);
}

/*
 * Called by the worker support thread to remove a worker
 * filesystem topology.
 *
 * NOTE: No need to conditionalize UseUsrSrc, it doesn't hurt to
 *	 issue the umount() if it isn't mounted and it ensures that
 *	 everything is unmounted properly on cleanup if the state
 *	 changes.
 */
void
DoWorkerUnmounts(worker_t *work)
{
	int retries;

	work->mount_error = 0;
	for (retries = 0; retries < 10; ++retries) {
		dounmount(work, "/proc");
		dounmount(work, "/dev");
		dounmount(work, "/usr/src");
		dounmount(work, "/usr/games");
		dounmount(work, "/boot/modules.local");
		dounmount(work, "/boot");
		dounmount(work, "/usr/local");
		dounmount(work, "/construction");
		dounmount(work, "/ccache");	/* in case of config change */
		dounmount(work, "/distfiles");
		dounmount(work, "/packages");
		dounmount(work, "/options");
		dounmount(work, "/xports");
		dounmount(work, "/usr/share");
		dounmount(work, "/usr/sbin");
		dounmount(work, "/usr/libexec");
		dounmount(work, "/usr/libdata");
		dounmount(work, "/usr/lib");
		dounmount(work, "/usr/include");
		dounmount(work, "/usr/bin");
		dounmount(work, "/libexec");
		dounmount(work, "/lib");
		dounmount(work, "/sbin");
		dounmount(work, "/bin");
		dounmount(work, "");
		if (work->mount_error == 0)
			break;
		sleep(5);
		work->mount_error = 0;
	}
	if (work->mount_error) {
		++work->accum_error;
		snprintf(work->status, sizeof(work->status),
			 "Unable to unmount slot");
	}
}

static
void
domount(worker_t *work, int type, const char *spath, const char *dpath,
	const char *discretefmt)
{
	const char *sbase;
	const char *rwstr;
	const char *optstr;
	const char *typestr;
	const char *debug;
	struct stat st;
	char *buf;
	char *tmp;
	int rc;

	/*
	 * Make target directory if necessary.  This must occur in-order
	 * since directories may have to be created under prior mounts
	 * in the sequence.
	 */
	asprintf(&buf, "%s%s", work->basedir, dpath);
	if (stat(buf, &st) != 0) {
		if (mkdir(buf, 0755) != 0) {
			fprintf(stderr, "Command failed: mkdir %s\n", buf);
			++work->mount_error;
		}
	}
	free(buf);

	/*
	 * Setup for mount arguments
	 */
	rwstr = (type & MOUNT_TYPE_RW) ? "rw" : "ro";
	optstr = "";
	typestr = "";

	switch(type & MOUNT_TYPE_MASK) {
	case MOUNT_TYPE_TMPFS:
		/*
		 * When creating a tmpfs filesystem, make sure the big ones
		 * requested are big enough for the worst-case dport (which
		 * is usually chromium).  If debugging is turned on, its even
		 * worse.  You'd better have enough swap!
		 */
		debug = getbuildenv("WITH_DEBUG");
		typestr = "tmpfs";
		if (type & MOUNT_TYPE_BIG)
			optstr = debug ? " -o size=128g" : " -o size=64g";
		else if (type & MOUNT_TYPE_MED)
			optstr = debug ? " -o size=32g" : " -o size=16g";
		else
			optstr = " -o size=16g";
		break;
	case MOUNT_TYPE_NULLFS:
#if defined(__DragonFly__)
		typestr = "null";
#else
		typestr = "nullfs";
#endif
		break;
	case MOUNT_TYPE_DEVFS:
		typestr = "devfs";
		break;
	case MOUNT_TYPE_PROCFS:
		typestr = "procfs";
		break;
	default:
		dfatal("Illegal mount type: %08x", type);
		/* NOT REACHED */
		break;
	}

	/*
	 * Prefix spath
	 */
	if (discretefmt) {
		sbase = BuildBase;
		asprintf(&tmp, discretefmt, work->index);
		spath = tmp;
	} else {
		if (spath[0] == '$') {
			++spath;
			sbase = SystemPath;
			if (strcmp(sbase, "/") == 0)
				++sbase;
		} else {
			sbase = "";
		}
		tmp = NULL;
	}
	asprintf(&buf, "%s%s -t %s -o %s %s%s %s%s",
		MOUNT_BINARY, optstr, typestr, rwstr,
		sbase, spath, work->basedir, dpath);
	rc = system(buf);
	if (rc) {
		fprintf(stderr, "Command failed: %s\n", buf);
		++work->mount_error;
	}
	free(buf);
	if (tmp)
		free(tmp);
}

static
void
dounmount(worker_t *work, const char *rpath)
{
	char *buf;

	asprintf(&buf, "%s%s", work->basedir, rpath);
	if (unmount(buf, 0) < 0) {
		switch(errno) {
		case EPERM:	/* This is probably fatal later on in mount */
		case ENOENT:	/* Expected if mount already gone */
		case EINVAL:	/* Expected if mount already gone (maybe) */
			break;
		default:
			fprintf(stderr, "Cannot umount %s (%s)\n",
				buf, strerror(errno));
			++work->mount_error;
			break;
		}
	}
	free(buf);
}

static
void
makeDiscreteCopies(const char *spath, const char *discretefmt)
{
	char *src;
	char *dst;
	char *buf;
	struct stat st;
	int i;
	int rc;

	for (i = 0; i < MaxWorkers; ++i) {
		if (spath[0] == '$') {
			if (strcmp(SystemPath, "/") == 0)
				asprintf(&src, "%s%s",
					 SystemPath + 1, spath + 1);
			else
				asprintf(&src, "%s%s",
					 SystemPath, spath + 1);
		} else {
			src = strdup(spath);
		}
		asprintf(&buf, discretefmt, i);
		asprintf(&dst, "%s%s", BuildBase, buf);
		free(buf);

		if (stat(dst, &st) < 0) {
			if (mkdir(dst, 0555) < 0) {
				dlog(DLOG_ALL, "Template - mkdir %s failed\n",
				     dst);
				dfatal_errno("Cannot mkdir %s:", dst);
			}
		}
		asprintf(&buf, "chflags -R noschg %s; "
			       "rm -rf %s; "
			       "cp -Rp %s/. %s",
			       dst, dst, src, dst);
		rc = system(buf);
		dlog(DLOG_ALL | DLOG_FILTER,
		     "Template - rc=%d running %s\n", rc, buf);
		if (rc)
			dfatal("Command failed: %s", buf);
		free(buf);
		free(src);
		free(dst);
	}
}
